src/animated/Controller.ts

```diff
@@ -69,123 +92,131 @@ class Controller<P extends any = {}> {
-    this.listeners = []
+    if (!this.idle || this.pendingCount) {
+      this.guid++
+      this._stop(finished)
+      this.pendingCount = 0
+    }
     return this
   }
 
-  /** Pause sets onEnd listeners free, but also removes the controller from the frameloop */
-  pause(finished?: boolean) {
-    this.stop(true)
-    if (finished) stop(this)
-    return this
+  // Called by the frameloop
+  onFrame(isActive: boolean) {
+    if (this.props.onFrame) {
+      this.props.onFrame(this.values)
+    }
+    if (!isActive) {
+      this._stop(true)
+    }
   }
 
-  runAsync({ delay, ...props }: P, onEnd: FinishedCallback) {
-    const local = this.local
+  runAsync(
+    guid: number,
+    props: SpringProps<DS>,
+    onEnd?: (finished: boolean) => void
+  ) {
     // If "to" is either a function or an array it will be processed async, therefor "to" should be empty right now
     // If the view relies on certain values "from" has to be present
-    let queue = Promise.resolve(undefined)
-    if (is.arr(props.to)) {
-      for (let i = 0; i < props.to.length; i++) {
-        const index = i
-        const fresh = { ...props, ...interpolateTo(props.to[index]) }
-        if (is.arr(fresh.config)) fresh.config = fresh.config[index]
-        queue = queue.then(
-          (): Promise<any> | void => {
-            //this.stop()
-            if (local === this.guid)
-              return new Promise(r => this.diff(fresh).start(r))
-          }
-        )
-      }
-    } else if (is.fun(props.to)) {
-      let index = 0
+    let queue = Promise.resolve()
+
+    const { to } = props
+    if (is.arr(to)) {
+      to.forEach((p, i) => {
+        queue = queue.then(() => {
+          if (guid !== this.guid) return
+          const fresh = { ...props, ...interpolateTo(p) }
+          if (is.arr(fresh.config)) fresh.config = fresh.config[i]
+          return new Promise<any>(r => this.diff(fresh).start(r))
+        })
+      })
+    } else if (is.fun(to)) {
+      let i = 0
       let last: Promise<any>
       queue = queue.then(() =>
-        props
-          .to(
-            // next(props)
-            (p: P) => {
-              const fresh = { ...props, ...interpolateTo(p) }
-              if (is.arr(fresh.config)) fresh.config = fresh.config[index]
-              index++
-              //this.stop()
-              if (local === this.guid)
-                return (last = new Promise(r => this.diff(fresh).start(r)))
-              return
-            },
-            // cancel()
-            (finished = true) => this.stop(finished)
-          )
-          .then(() => last)
+        to(
+          // next(props)
+          p => {
+            if (guid !== this.guid) return
+            const fresh = { ...props, ...interpolateTo(p) }
+            if (is.arr(fresh.config)) fresh.config = fresh.config[i++]
+            return (last = new Promise(r => this.diff(fresh).start(r)))
+          },
+          // cancel()
+          finished => this.stop(finished)
```

Comment:
The `finished` argument used to default to `true`, but that didn't make sense to me. Usually, if you're calling this function, it's to signal that the animation is cancelled. To signal that the animation finished cleanly, just return without calling either of these functions.

Reply:
the question is what constitutes an end, it could be merely that listeners are fired, or that the controller is ripped out of the frameloop. both have usecases with side effects in some specific situations.

Reply:
I think of an "animation end" as (1) a spring/decay coming to rest, (2) a duration hitting zero, (3) the last keyframe ending, or (4) an async animator returning undefined. The frameloop should be an implementation detail whenever possible. Does that clear things up?

Reply:
Makes sense!